File: 301-400/0354-Max-Envelopes/0354.go
func max(a ...int) int {
	res := a[0]
	for _, v := range a[1:] {
		if v > res {
			res = v
		}
	}
	return res
}

func maxEnvelopes(envelopes [][]int) int {
	n := len(envelopes)
	if n == 0 {
		return 0
	}

	sort.Slice(envelopes, func(i, j int) bool {
		a, b := envelopes[i], envelopes[j]
		return a[0] < b[0] || a[0] == b[0] && a[1] > b[1]
	})

	f := make([]int, n)
	for i := range f {
		f[i] = 1
	}
	for i := 1; i < n; i++ {
		for j := 0; j < i; j++ {
			if envelopes[j][1] < envelopes[i][1] {
				f[i] = max(f[i], f[j]+1)
			}
		}
	}
	return max(f...)
}

File: 301-400/0354-Max-Envelopes/0354.py
class Solution:
  def maxEnvelopes(self, envelopes: List[List[int]]) -> int:
    if not envelopes:
      return 0

    n = len(envelopes)
    envelopes.sort(key=lambda x: (x[0], -x[1]))

    f=  [1] * n
    for i in range(n):
      for j in range(i):
        if (envelopes[j][1] < envelopes[i][1]):
          f[i] = max(f[i], f[j] + 1)

    return max(f)